### PR TITLE
[Proposal] Magic rework (Phase 1)

### PR DESCRIPTION
**Pull request by Xaver-DaRed**
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] that I agree to LandSandBoat's [Limited Contributor License Agreement](https://github.com/LandSandBoat/server/blob/base/.github/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I have **read the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)**
- [eh...] that I've _**tested my code and things my code changed**_ since the last commit in the PR, and will test after any later commits

An atempt to make magic calculations sane, to put it simply.

- What this is:

Right now, spells are being handled in magic.lua, core and their own spell scripts.
This would be fine, but magic.lua has many problems.
The first being that it is very hard to follow, with the effect of making a correction to spells a lot harder than it should be.
The second is that it handles calculations chaotically and outright wrong, making magic not work at all as it should.

This attempts to make magic logic easier to follow by dividing the different magic processes into their own scripts, stripping logic from spell scripts in the process.

This also atempts to replicate retails damage calculations better than how we currently do it.

This is only the first phase, wich handles damage calculations. Second phase would handle enfeebling spells, and further phases other types of spells (healing, enhancing, etc...)

- What this isn't:

A replacement for magic.lua. Spells affected don't require it, but spells unaffected do.
The endgame would be not to need it,

A fix for every problem. Or for any problem. If it fixes something is by sheer coincidence/luck/mistake.

A finished product.

A fully tested product. I'm just one person. It's tested enough to know it compiles, it doesn't IMEDIATELY crash the server and spells do actual damage.

### CODE REVIEW

File: scripts/globals/magic_utils/spell_damage.lua [hunk 1]
```diff
@@ -0,0 +1,863 @@
...
+function calculateBaseDamage(caster, target, spell, spellId, skillType, statDiff)
+    local spellDamage          = 0 -- The variable we want to calculate
+    local baseSpellDamage      = 0 -- (V) In Wiki.
+    local baseSpellDamageBonus = 0 -- (mDMG) In Wiki. Get from equipment, status, etc...
+    local statDiffBonus        = 0 -- statDiff x apropiate multipliers.
+
+    -- Get base damage (V in English wiki, D in JP wiki) depending on caster.
+    if caster:isPC() then
+        baseSpellDamage = damageSpellParams[spellId][6]
+    else
```
Comment (claywar) [inline scripts/globals/magic_utils/spell_damage.lua:70]: Just a suggestion, might want to define some constants to cover the magic numbers in most table references.  The arrays are pretty beefy, and that would go a long way for readability (this continues throughout the PR)

File: scripts/globals/magic_utils/spell_damage.lua [hunk 2]
```diff
@@ -0,0 +1,863 @@
...
+            statDiffBonus = math.floor(50 * spellMultiplier0) + math.floor(50 * spellMultiplier50) + math.floor(100 * spellMultiplier100) + math.floor((statDiff - 200) * spellMultiplier200)
+        elseif statDiff < 400 and statDiff >= 300 then
+            statDiffBonus = math.floor(50 * spellMultiplier0) + math.floor(50 * spellMultiplier50) + math.floor(100 * spellMultiplier100) + math.floor(200 * spellMultiplier200) + math.floor((statDiff - 300) * spellMultiplier300)
+        elseif statDiff < 500 and statDiff >= 400 then
+            statDiffBonus = math.floor(50 * spellMultiplier0) + math.floor(50 * spellMultiplier50) + math.floor(100 * spellMultiplier100) + math.floor(200 * spellMultiplier200) + math.floor(300 * spellMultiplier300) + math.floor((statDiff - 400) * spellMultiplier400)
+        else -- It's over 500!
+            statDiffBonus = math.floor(50 * spellMultiplier0) + math.floor(50 * spellMultiplier50) + math.floor(100 * spellMultiplier100) + math.floor(200 * spellMultiplier200) + math.floor(300 * spellMultiplier300) + math.floor(400 * spellMultiplier400) +  math.floor((statDiff - 500) * spellMultiplier500)
+        end
```
Comment (claywar) [inline scripts/globals/magic_utils/spell_damage.lua:108]: Can we handle some of this through multiple lines?

File: scripts/globals/magic_utils/spell_damage.lua [hunk 3]
```diff
@@ -0,0 +1,863 @@
...
+        local mobSpellMultiplier = damageSpellParams[spellId][5] -- (M) In wiki. Mob value in BG wiki.
+        statDiffBonus = statDiff * mobSpellMultiplier
+    end
+
+    -- Calculate baseSpellDamageBonus. Player only.
+    if caster:isPC() then
+        -- BLM Job Point: Manafont Elemental Magic Damage +3
+        if caster:hasStatusEffect(xi.effect.MANAFONT) then
+            baseSpellDamageBonus = baseSpellDamageBonus + (caster:getJobPointLevel(xi.jp.MANAFONT_EFFECT) * 3)
+        end
```
Comment (claywar) [inline scripts/globals/magic_utils/spell_damage.lua:136]: Should be fine with order of ops without the parens

File: scripts/globals/magic_utils/spell_damage.lua [hunk 4]
```diff
@@ -0,0 +1,863 @@
...
+    -- No negative base damage value allowed.
+    if spellDamage < 0 then
+        spellDamage = 0
+    end
+
+    return spellDamage
+end
+
+function calculateMTDR(caster, target, spell)
+    local MTDR    = 0 -- The variable we want to calculate.
```
Comment (claywar) [inline scripts/globals/magic_utils/spell_damage.lua]: Can there be a comment defining what MTDR is, or perhaps a change in naming convention?

File: scripts/globals/magic_utils/spell_damage.lua [hunk 5]
```diff
@@ -0,0 +1,863 @@
...
+
+    return MTDR
+end
+
+function calculateEleStaffBonus(caster, spell, spellElement)
+    local eleStaffBonus = caster:getMod(strongAffinityDmg[spellElement])
+
+    if eleStaffBonus > 0 then
+        eleStaffBonus = 1 + (eleStaffBonus * 0.05)
+    else
```
Comment (claywar) [inline scripts/globals/magic_utils/spell_damage.lua:189]: Unneeded parens

File: scripts/globals/magic_utils/spell_damage.lua [hunk 6]
```diff
@@ -0,0 +1,863 @@
...
+        if skillchainCount > 0 then
+            magicAcc = magicAcc + 25
+        end
+
+        -----------------------------------
+        -- magicAcc from Job Points.
+        -----------------------------------
+        -- WHM Job Points
+        if casterJob == xi.job.WHM then
+            magicAcc = magicAcc + caster:getJobPointLevel(xi.jp.WHM_MAGIC_ACC_BONUS)
```
Comment (claywar) [inline scripts/globals/magic_utils/spell_damage.lua]: This is more on me for the initial changes, but these can move to if/elseif, since mainjob won't be changing during cast

File: scripts/globals/magic_utils/spell_damage.lua [hunk 7]
```diff
@@ -0,0 +1,863 @@
...
+            then
+                magicAcc = magicAcc + caster:getJobPointLevel(xi.jp.STRATEGEM_EFFECT_I)
+            end
+        end
+
+        -----------------------------------
+        -- magicAcc from Merits.
+        -----------------------------------
+        -- BLM Merits
+        if casterJob == xi.job.BLM and skillType == xi.skill.ELEMENTAL_MAGIC then
```
Comment (claywar) [inline scripts/globals/magic_utils/spell_damage.lua]: Same as above for job checks

File: scripts/globals/magic_utils/spell_damage.lua [hunk 8]
```diff
@@ -0,0 +1,863 @@
...
+    end
+
+    -- Calculate Weather bonus
+    if weather == xi.magic.singleWeatherStrong[spellElement] then
+        if caster:getMod(xi.mod.IRIDESCENCE) >= 1 then
+            if math.random() < 0.33 or caster:getMod(elementalObi[spellElement]) >= 1 or isHelixSpell == true then
+                dayAndWeather = dayAndWeather + 0.10
+            end
+        end
+        if math.random() < 0.33 or caster:getMod(elementalObi[spellElement]) >= 1 or isHelixSpell == true then
```
Comment (claywar) [inline scripts/globals/magic_utils/spell_damage.lua]: Small styling preference on my end, if there's multiple conditionals, could you put a newline in between to easily differentiate between blocks?

File: scripts/globals/magic_utils/spell_damage.lua [hunk 9]
```diff
@@ -0,0 +1,863 @@
...
+
+    if magicBonusDiff < 0 then
+        magicBonusDiff = 0
+    end
+
+    return magicBonusDiff
+end
+
+function calculateTMDA(caster, target, spell, spellElement)
+    local TMDA = 1 -- The variable we want to calculate
```
Comment (claywar) [inline scripts/globals/magic_utils/spell_damage.lua:609]: Same with MTDR, TMDA could use a comment or naming that identifies what its purpose is

File: scripts/globals/magic_utils/spell_damage.lua (hunk 4 above)
Comment (Xaver-DaRed) [reply scripts/globals/magic_utils/spell_damage.lua]: I followed BG wiki naming conventions, but you are right, I'll add a comment.
MTDR stands for "Multiple Target Damage Reduction" btw XD

File: scripts/globals/magic_utils/spell_damage.lua [hunk 10]
```diff
@@ -0,0 +1,883 @@
...
+        -- NIN Merits
+        elseif casterJob == xi.job.NIN and skillType == xi.skill.NINJUTSU then
+            magicAcc = magicAcc + caster:getMerit(xi.merit.NIN_MAGIC_ACCURACY)
+
+        -- BLU Merits
+        elseif casterJob == xi.job.BLU and skillType == xi.skill.BLUE_MAGIC then
+            magicAcc = magicAcc + caster:getMerit(xi.merit.MAGICAL_ACCURACY)
+        end
+        -- OTHER merits. ???? which are...? Modifiers need a rework/reorganization
+        magicAcc = magicAcc + caster:getMerit(xi.merit.MAGIC_ACCURACY)
```
Comment (claywar) [inline scripts/globals/magic_utils/spell_damage.lua]: Where is this generic `xi.merit.MAGIC_ACCURACY` coming from?

Comment (Xaver-DaRed) [reply scripts/globals/magic_utils/spell_damage.lua]: As I pointed out in the comment, I have no clue XD
It was there in the original magic.lua file and decided to keep it in case someone had more info on it.
It is on my "to destroy if no info is provided" private list

File: scripts/globals/magic_utils/spell_damage.lua [hunk 11]
```diff
@@ -0,0 +1,883 @@
...
+    [xi.magic.spell.WATER_VI    ] = { xi.mod.INT, xi.skill.ELEMENTAL_MAGIC,     xi.magic.element.WATER, 1010,  1.5, 1010,  6.5,  5.9,  4.9,   3.9, 2.95, 1.99,    1 },
+    [xi.magic.spell.FLOOD       ] = { xi.mod.INT, xi.skill.ELEMENTAL_MAGIC,     xi.magic.element.WATER,  552,    2,  700,    2,    2,    2,     2,    2,    2,    2 },
+    [xi.magic.spell.FLOOD_II    ] = { xi.mod.INT, xi.skill.ELEMENTAL_MAGIC,     xi.magic.element.WATER,  710,    2,  800,    2,    2,    2,     2,    2,    2,    2 },
+    [xi.magic.spell.COMET       ] = { xi.mod.INT, xi.skill.ELEMENTAL_MAGIC,      xi.magic.element.DARK,  964,  2.3, 1000,    4, 3.75,  3.5,     3,    2,    1,    1 },
+
+-- Multiple target spells:
+-- Structure:           [spellId] = {  Stat used,                    Skill,                    Element, vNPC, mNPC,  vPC,   M0,  M50,  M100, M200, M300, M400, M500 },
+    [xi.magic.spell.AEROGA      ] = { xi.mod.INT, xi.skill.ELEMENTAL_MAGIC,      xi.magic.element.WIND,   93,    1,  100,  2.6,  1.8,     1,    0,    0,    0,    0 },
```
Comment (TeoTwawki) [inline scripts/globals/magic_utils/spell_damage.lua]: Maybe we could make these liens less wide by casting some things to variables or by getting some things from other locations rather than having to repeat them into the table/array

like
```lua
local INT = xi.mod.INT
local elemSkill = xi.skill.ELEMENTAL_MAGIC
```
```lua
[xi.magic.spell.AEROGA      ] = { INT, elemSklill,      xi.magic.element.WIND,   93,    1,  100,  2.6,  1.8,     1,    0,    0,    0,    0 },
```
or we use `spell:getSkillType()` to not need a table entry at all

Comment (TeoTwawki) [reply scripts/globals/magic_utils/spell_damage.lua]: we can "get" the element as well.

Comment (Xaver-DaRed) [reply scripts/globals/magic_utils/spell_damage.lua]: The reason I defined elements (and skills) into the table was to make less calls to external functions. I'm guessing it is making a call to the database.
I was hoping that someone (not me, someone that actually knows how to do it) would compare both methods and check wich one was more eficient resource wise.
It also helped me when filling the tables. A visual help, so to speak.

I'm not oposed to creating local vars to shorten the length though.

**Comment by zach2good** [comment]
I don't have time to look at this for a few days, but a 90-second high level skim leaves me very impressed. I thought I would hate the giant table of spell parameters but it looks pretty good!

I'll try to get a review in at the weekend 👍 

I don't know if this would make the information too dense, but you can reduce how wide your tables are like this:
```lua
-- Assign a shorter name
local ma = xi.magic
local we = xi.weather
local ef = xi.effect
local mod = xi.mod

-- Big structure table
```

This will work as long as the values aren't being written to, I'm not sure if this operates by value or reference 🤷

File: scripts/globals/magic_utils/spell_damage.lua (hunk 11 above)
Comment (zach2good) [reply scripts/globals/magic_utils/spell_damage.lua]: I see Teo has already had the same idea as me...

**Comment by zach2good** [comment]
One of the strengths of this design is how easy it is for people who want to use era-style values/calculations.
If the big tables of params were changes to be attached to the `xi` global:

```lua
local damageSpellParams = ...
```
->
```lua
xi.magic_utils.damageSpellParams = ...
```

`modules/era_spell_values.lua` ->
```lua
local module = require("scripts/modules/module")
module.override(xi.magic_utils, "damageSpellParams", {
    -- Table of era-style values
}
)
```

... or `spell_damage` can be overridden,

**Comment by TeoTwawki** [comment]
> 
> 
> One of the strengths of this design is how easy it is for people who want to use era-style values/calculations.
> If the big tables of params were changes to be attached to the `xi` global:
> 
> ```lua
> local damageSpellParams = ...
> ```
> 
> ->
> 
> ```lua
> xi.magic_utils.damageSpellParams = ...
> ```
> 
> `modules/era_spell_values.lua` ->
> 
> ```lua
> local module = require("scripts/modules/module")
> module.override(xi.magic_utils, "damageSpellParams", {
>     -- Table of era-style values
> }
> )
> ```
> 
> ... or `spell_damage` can be overridden,

its also turns out that monsters are still using those era calculations too so now we have a retail accuracy reason to continue supporting them

File: scripts/globals/magic_utils/spell_damage.lua (hunk 10 above)
Comment (TeoTwawki) [reply scripts/globals/magic_utils/spell_damage.lua]: its a modifier set by reading the [merit table](https://github.com/LandSandBoat/server/blob/base/sql/merits.sql#L226). Whether or not you can [actually merit it ](https://github.com/LandSandBoat/server/blob/base/sql/char_merit.sql)is another story.

The code fr applying merits is rather generic so you won't find core code doing application of the specific merit.

File: scripts/globals/magic_utils/spell_damage.lua (hunk 8 above)
Comment (TeoTwawki) [reply scripts/globals/magic_utils/spell_damage.lua]:
```lua
if
    math.random() < 0.33 or
    caster:getMod(elementalObi[spellElement]) >= 1 or
    isHelixSpell == true
then
```
is the format specified by the style guide when conditionals are to wide or confusing, iirc

ps we usual do not need `== true` for most things, unless they can be something that is not true/false (nil?)

File: scripts/globals/magic_utils/spell_damage.lua (hunk 10 above)
Comment (Xaver-DaRed) [reply scripts/globals/magic_utils/spell_damage.lua]: oh, I see. It's RDM category 2 merit

File: scripts/globals/magic_utils/spell_damage.lua (hunk 8 above)
Comment (Xaver-DaRed) [reply scripts/globals/magic_utils/spell_damage.lua]: Ya, I didnt consider it too long, honestly. That + the repetition in the following lines made me think it was ok as it is...
If there were both "or" and "and", I would have done it already.
If you both agree to divide this (and the ones bellow by extension), I will do it, but i think it will only serve to bloat this particular chain.

File: scripts/globals/spells/waterga_v.lua [hunk 12]
```diff
@@ -0,0 +1,21 @@
+-----------------------------------
+-- Spell: Waterga V
+-----------------------------------
+require("scripts/globals/magic_utils/spell_damage")
+require("scripts/globals/msg")
+-----------------------------------
+local spell_object = {}
+
+spell_object.onMagicCastingCheck = function(caster, target, spell)
+    if caster:isPC() then -- Mob Only Spell.
```
Comment (TeoTwawki) [inline scripts/globals/spells/waterga_v.lua]: this isn't necessary. a player normally can't obtain it to cast it, we do not need to block it. If you choose to GM yourself the spell to get it, its not a bug its just not the projects problem either.

File: scripts/globals/magic_utils/spell_damage.lua [hunk 13]
```diff
@@ -0,0 +1,908 @@
...
+
+    --     return magicBonusDiff
+    -- end
+
+    -- Ninja spell bonuses
+    if skillType == xi.skill.NINJUTSU then
+        -- Ninja Category 2 merits.
+        mab = mab + caster:getMerit(xi.merit.NIN_MAGIC_BONUS)
+        -- Ninja Category 1 merits
+        if spellId >= 320 and spellId <= 322 then     -- Katon series.
```
Comment (zach2good) [inline scripts/globals/magic_utils/spell_damage.lua]: You should be able to use `xi.magic.spellFamily` for these checks:
https://github.com/LandSandBoat/server/blob/base/scripts/globals/spell_data.lua#L113

For checks (like ancient magic) I wouldn't be against helpers living somewhere so you can just call `xi.spell.isAncientMagic(spellId)`. Even if you've commented it well

File: scripts/globals/magic_utils/spell_damage.lua (hunk 1 above)
Comment (TeoTwawki) [reply scripts/globals/magic_utils/spell_damage.lua:70]: like
```lua
local vPC = 4
--[[ stuff between]]
baseSpellDamage = damageSpellParams[spellId][vPC]
```
?

Comment (Xaver-DaRed) [reply scripts/globals/magic_utils/spell_damage.lua:70]: yeah, thats exactly it.
Gonna do it RQ ans see what the new lua-checks return while im at it

File: scripts/globals/magic_utils/spell_damage.lua (hunk 8 above)
Comment (TeoTwawki) [reply scripts/globals/magic_utils/spell_damage.lua]: I think I misread what clay was asking for here before quoting styling info. I think he just wanted a blank line between the previous `end` and the next `if`.

Comment (Xaver-DaRed) [reply scripts/globals/magic_utils/spell_damage.lua]: oh. I see... kk, easy enough

**Comment by Xaver-DaRed** [comment]
Ok, thats gonna need to be squashed...

Finished fixing most warnings given by CI.
Warnings in spell scripts are out of the scope of this PR. Mainly becouse they involve phase 2 anyway (rewrite debuff logic)
Warnings in main file... well, maybe i could do something about those, but honestly, they involve a different file altoguether and If I dare to open it, I would have to rewrite it too.
I refuse to, at this time XD

**Comment by kaincenteno** [comment]
Love the param table, so easy to read than going through each spell. Def an improvement :smile:

File: scripts/globals/magic_utils/spell_damage.lua (hunk 13 above)
Comment (ShiyoKozuki) [reply scripts/globals/magic_utils/spell_damage.lua:558]: Few things:
**Elemental Seal:**
From personal testing(No proof, sorry), elemental seal seems to make your resist go up a tier. So if you have 50% SDT, it goes to from 1/2 to 1. If you have sub 50% SDT, it goes from 1/4 to 1/2. 

**Resist stats for players only**
I have testing for this and proof!
https://www.ffxiah.com/forum/topic/46016/first-and-final-line-of-defense-v20/107/#3570888

https://www.ffxiah.com/forum/topic/46016/first-and-final-line-of-defense-v20/109/#3571304

```
Enfeebling test on a 2 resist state enfeeble, via having level ~70 Colibris reflect Dispel. I had Saboteur Addle II on the mob and 4/5 Malignance, so MEVA should be capped.

Neutral dark resistance: 1 landed, 9 resisted
-3 dark resistance: 10 landed
```

So, basically..

If you have any negative resists to the element hitting you, it will always land for 1/2 resist minimum(including enfeebles)
If you do not have any positive resist to the element hitting you, you cannot resist to 1/8.

**Comment by ShiyoKozuki** [comment]
Few things.

**Elemental Sea**l:
From personal testing(No proof, sorry), elemental seal seems to make your resist go up a tier. So if your enemy has 50% SDT, it goes to from 1/2 to 1. If your enemy has sub 50% SDT, it goes from 1/4 to 1/2. Finally, if your enemy has 5% SDT, it goes from 1/8th to 1/4th. It's also guaranteed to land NO MATTER WHAT except for status resist trait procs.
_Requires further testing._

**Resist stats for players only**:
I have testing for this and proof!
https://www.ffxiah.com/forum/topic/46016/first-and-final-line-of-defense-v20/107/#3570888

https://www.ffxiah.com/forum/topic/46016/first-and-final-line-of-defense-v20/109/#3571304

```
Enfeebling test on a 2 resist state enfeeble, via having level ~70 Colibris reflect Dispel. I had Saboteur Addle II on the mob and 4/5 Malignance, so MEVA should be capped.

Neutral dark resistance: 1 landed, 9 resisted
-3 dark resistance: 10 landed
```
So, basically..

If you have any negative resists to the element hitting you, it will always land for 1/2 resist minimum(including enfeebles)
If you do not have any positive resist to the element hitting you, you cannot resist to 1/8.


**Drain / Aspir**

Limited retail testing seems to indicate they get a flat end multiplier based on SDT. Aka, drain/aspir damage * SDT. So 150% SDT would be drain/aspir final damage * 1.5 while 50% SDT would be drain/aspir final damage * 0.5
_Requires further testing._

**Stun / Flash  / Dispel(Light and dark variants)**
They all have some type of varying macc bonus. Stun from /drk on a +15 monster with neutral SDT would land 100% of the time at 75. When I played SCH at 99 cap(not ilvl), I needed obscene macc to land stun reliably on Legion enemies with high SDT to thunder and high mevasion.
_Requires further testing._

**Comment by TeoTwawki** [comment]
> When I played SCH at 99 cap(not ilvl), I needed obscene macc to land stun reliably on Legion enemies with high SDT to thunder and high mevasion.

you should know that Legion had modified stun resistances all over creation because some of the admins didn't want it being too reliable to "just throw more blm or blue at". drk was collateral damage. A lot of (but not all of) that nonsense was stripped out (by me) prior to their public code dump when they shut down.


not saying there isn't anything else, but thought was worth the mention/grain of salt on using legion as the example.

File: scripts/globals/magic_utils/spell_damage.lua [hunk 14]
```diff
@@ -0,0 +1,790 @@
...
+            if math.random() < 0.33 or caster:getMod(elementalObi[spellElement]) >= 1 or isHelixSpell then
+                dayAndWeather = dayAndWeather + 0.10
+            end
+        end
+
+        if math.random() < 0.33 or caster:getMod(elementalObi[spellElement]) >= 1 or isHelixSpell then
+            dayAndWeather = dayAndWeather + 0.10
+        end
+
+    elseif weather == xi.magic.singleWeatherWeak[spellElement] then
```
Comment (TeoTwawki) [inline scripts/globals/magic_utils/spell_damage.lua]: possibly to table/loop much if this if/else tower? it looks like a lot of repeating code goes on here.

File: scripts/globals/magic_utils/spell_damage.lua [hunk 15]
```diff
@@ -0,0 +1,790 @@
...
+    return magicBurst
+end
+
+xi.magic_utils.spell_damage.calculateIfMagicBurstBonus = function(caster, target, spell, spellId, spellElement)
+    local magicBurstBonus                 = 1.0 -- The variable we want to calculate
+    local modBurst                        = 1.0
+    local ancientMagicBurstBonus          = 0
+    local skillchainTier, skillchainCount = FormMagicBurst(spellElement, target) -- External function. Not present in magic.lua.
+
+    if spellId >= 204 and spellId <= 215 then -- If spell is Ancient Magic
```
Comment (TeoTwawki) [inline scripts/globals/magic_utils/spell_damage.lua]: I think we have spell names enum'd somewhere now so that we don't have to have the ID use

**Comment by ShiyoKozuki** [comment]
> > When I played SCH at 99 cap(not ilvl), I needed obscene macc to land stun reliably on Legion enemies with high SDT to thunder and high mevasion.
> 
> you should know that Legion had modified stun resistances all over creation because some of the admins didn't want it being too reliable to "just throw more blm or blue at". drk was collateral damage. A lot of (but not all of) that nonsense was stripped out (by me) prior to their public code dump when they shut down.
> 
> not saying there isn't anything else, but thought was worth the mention/grain of salt on using legion as the example.

Legion meaning the end game event, not the server.

File: scripts/globals/magic_utils/spell_damage.lua [hunk 16]
```diff
@@ -43,8 +43,7 @@ local blmMerit               = { xi.merit.FIRE_MAGIC_POTENCY,  xi.merit.ICE_MAGI
 local rdmMerit               = { xi.merit.FIRE_MAGIC_ACCURACY, xi.merit.ICE_MAGIC_ACCURACY, xi.merit.WIND_MAGIC_ACCURACY,  xi.merit.EARTH_MAGIC_ACCURACY, xi.merit.LIGHTNING_MAGIC_ACCURACY, xi.merit.WATER_MAGIC_ACCURACY   }
 
 -- Table variables.
-local spellTable   = xi.magic_utils.parameters.damageParams
-local elementTable = xi.magic_utils.parameters.elementParams
+local table = xi.magic_utils.parameters.damageParams
```
Comment (TeoTwawki) [inline scripts/globals/magic_utils/spell_damage.lua]: bad variable name, avoid using keywords as names

**Comment by Xaver-DaRed** [comment]
Went through and tested this.
- Nothing crashes or errors IMMEDIATELY.
- Magic Bursts work and modifiers are applied.
- Resists happen.
- SDT seems to happen aswell.

File: scripts/globals/magic_utils/spell_damage.lua [hunk 17]
```diff
@@ -206,7 +206,10 @@ end
 -- Careful description required to prevent further problems where people misunderstand what is happening.
 -- Multiple things can affect dmg. in our core SDT is a % reduction in its own step.
 xi.magic_utils.spell_damage.calculateSDT = function(caster, target, spell, spellElement)
-    local SDT = target:getMod(xi.magic.specificDmgTakenMod[spellElement]) -- The variable we want to calculate
+    local SDT    = 1 -- The variable we want to calculate
+    local SDTMod = target:getMod(xi.magic.specificDmgTakenMod[spellElement]) -- Elemental SDT mods are Base 100.
+
+    SDT = (SDTMod / -100) + 1 -- Convert Modifier value.
```
Comment (TeoTwawki) [inline scripts/globals/magic_utils/spell_damage.lua]: ~~like I was saying on discord, the sql has to do that to create modifiers. you do not want to repeat/reverse it and they are -not- base 100 they are base 1000 to support half percent~~

From modifier.h and status.lua:
```
Example: 1000 = 100%, 875= 87.5%
```
~~this confusion is precisely why I left that comment there~~

Just realized everything wasn't fully converted script side. will take care of this when it is

**Review by TeoTwawki — APPROVED** [review]
Outside of an expected SDT [issue](https://github.com/LandSandBoat/server/issues/1046) we can handle after, this looks good to me.

**Review by Era-Lusiphur — APPROVED** [review]
Did a good bit of before/after testing, looked at results with @TeoTwawki. Everything appears to have results with an expected SDT skew Teo was anticipating. 

Tested blue magic, divine, elemental, enfeebling, and songs.
